os/kernel/fault_manager/fault_manager.c

```diff
@@ -0,0 +1,291 @@
+	}
+	printf("-------------------------------------\n");
+
+}
+
+/****************************************************************************
+ * Name: get_appid
+ *
+ * Description:
+ *   This function will traverse the task/thread tree and return the root app pid.
+ *   If fault happened in any of the thread/task, it will call this function to get
+ *   root application pid for the same.
+ *
+ * Input parameters:
+ *   pid   -   The pid of the task/thread, whose root app pid is required
+ *
+ * Returned Value:
+ *   pid   -   pid of the root application of input pid
+ *
+ ****************************************************************************/
+pid_t get_appid(pid_t pid)
+{
+	/* TODO: Implementation of traverser of the tree to get root app pid */
+	return pid;
+}
+
+/****************************************************************************
+ * Name: kill_app
+ *
+ * Description:
+ *   This function will kill all the task/thread created by the root application
+ *   i.e input pid.
+ *   If any fault occurs in the system, fault manager will call this function to
+ *   kill all the thread/task created by the input application.
+ *
+ * Input parameters:
+ *   pid   -   The pid of the applciation, whoes all childs to be killed
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ ****************************************************************************/
+int kill_app(pid_t pid)
+{
+	int ret;
+
+	/* TODO: Implementation of traverser of tree and kill all the task/thread
+	 * of the root application
+	 */
+	ret = kill(pid, SIGWORK);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: fault_manager
+ *
+ * Description:
+ *   This function will receive the faulty application's pid and check if it is one
+ *   of the registered application with fault manager.
+ *   If the application is registered then it will be restarted after killing its
+ *   child processes, and if it is not registered then board will be rebooted.
+ *
+ * Input parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ * Assumptions:
+ *    All the applications of the system will be registered with the fault manager
+ *
+ ****************************************************************************/
+int fault_manager(int argc, char *argv[])
+{
+	mqd_t mqdes;
+	int ret;
+	int i;
+	int index = -1;
+	pid_t pid;
+	pid_t pid_restart;
+	pid_t app_id;
+	struct mq_attr attr;
+	char restart_app[PATH_LEN] = " ";
+
+	attr.mq_maxmsg = 32;
+	attr.mq_msgsize = sizeof(pid);
+	attr.mq_flags = 0;
+
+	mqdes = mq_open(FAULT_MGR_MQ, O_CREAT | O_RDONLY, 0666, &attr);
+	DEBUGASSERT(mqdes != (mqd_t)-1);
+
+	while (1) {
+		sleep(2);
+
+		/* Wait for the message from assert handler for the faulty app pid */
+		ret = mq_receive(mqdes, (char *)&pid, sizeof(pid), NULL);
+		DEBUGASSERT(ret != -1);
+		DEBUGASSERT(ret == sizeof(pid));
```

Comment:
Both can be combined in a single statement.

Reply:
okay, I will modify this.
Thanks.